src/run/cluster.cpp

```diff
@@ -151,10 +244,273 @@ void run() {
+	// LazyDisjointIntegralSet<uint32_t> test(5);
+	// test.merge(1,3);
+	// test.merge(3,4);
+	// test.merge(0,2);
+	// auto los = test.getListOfSets();
+	// for(auto& set: los){
+	// 	for(auto item:set){
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+	// unordered_set<string> s ( {"red","green","blue","rain", "sun", "snow"} );
+	// LazyDisjointTypeSet<string> ts(&s);
+
+	// ts.merge("red", "green");
+	// ts.merge("snow", "rain");
+	// ts.merge("sun", "rain");
+	// auto lo = ts.getListOfSets();
+	// for(auto& set: lo){
+	// 	cout<< "set :";
+	// 	for(auto item:set){
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+
+	const size_t seq_count = db.ref_header.sequences;
+	statistics.reset();
+	config.command = Config::blastp;
+	config.no_self_hits = true;
+	//config.output_format = { "6", "qnum", "snum" };
+	config.output_format = { "6", "qnum", "snum", "qcovhsp", "scovhsp", "bitscore" };
+	config.query_cover = 80;
+	config.subject_cover = 80;
+	config.algo = 0;
+	config.index_mode = 0;
+	config.freq_sd = 0;
+	config.max_alignments = numeric_limits<uint64_t>::max();
+
+	Workflow::Search::Options opt;
+	opt.db = &db;
+	opt.self = true;
+	NeighborStream ns(db.ref_header.sequences);
+	opt.consumer = &ns;
+	opt.db_filter = nullptr;
+
+	Workflow::Search::run(opt);
+
+	auto clusterResult = ns.getListOfSets();
+	cout << "Found "<<clusterResult.size()<<" clusters"<<endl;
+	// for(auto& set: clusterResult){
+	// 	cout<< "set :";
+	// 	for(auto item : set){
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+}
+
+Eigen::SparseMatrix<float> get_gamma(Eigen::SparseMatrix<float>* m, float r){
+	vector<Eigen::Triplet<float>> data;
+	for (uint32_t k=0; k<m->outerSize(); ++k){
+		float colSum = 0.0f;
+		for (Eigen::SparseMatrix<float>::InnerIterator it(*m, k); it; ++it){
+			colSum += pow(it.value(),r);
+		}
+		for (Eigen::SparseMatrix<float>::InnerIterator it(*m, k); it; ++it){
+			data.emplace_back(it.row(), it.col(), pow(it.value(),r) / colSum);
+		}
+	}
+	Eigen::SparseMatrix<float> gamma(m->rows(), m->cols());
+	gamma.setFromTriplets(data.begin(), data.end());
+	return gamma.pruned(1.0, std::numeric_limits<float>::epsilon());
+}
+
+vector<unordered_set<uint32_t>> get_list(Eigen::SparseMatrix<float>* m){
+	LazyDisjointIntegralSet<uint32_t> disjointSet(m->cols());
+	for (uint32_t k=0; k<m->outerSize(); ++k){
+		for (Eigen::SparseMatrix<float>::InnerIterator it(*m, k); it; ++it){
+			disjointSet.merge(it.row(), it.col());
+		}
+	}
+	return disjointSet.getListOfSets();
+}
+
+vector<unordered_set<uint32_t>> markov_process(Eigen::SparseMatrix<float>* m){
+	uint32_t iteration = 0;
+	double diff_norm = std::numeric_limits<double>::max();
+	*m  = get_gamma(m,1); // This is to get a matrix of random walks on the graph -> TODO: find out if something else is more suitable
+	while( iteration < 100 && diff_norm > 1e-6*m->rows() ){
+		Eigen::SparseMatrix<float> msquared = ((*m) * (*m)).pruned(1.0, std::numeric_limits<float>::epsilon());
```

Comment:
This could probably be optimized by avoiding some memory allocations for new matrices, see here: http://eigen.tuxfamily.org/index.php?title=FAQ#Where_in_my_program_are_temporary_objects_created.3F

Reply:
Possibly. I am not sure how to supply Eigen with the required memory for the temporary object though. I mean the memory requirements throughout the iterations are pretty static and it would be nice to just reuse the allocated space (and I hoped that Eigen does that under the hood). Do you have any reference on how to do this (if not, I can try and google it myself ;) )

Reply:
The idea is basically to avoid the creation of temporarys, something like this:

```
Eigen::SparseMatrix<float> squared, diff, update;
while(...) {
  squared = m;
  squared *= m;
  get_gamma(squared, update)
  diff = m;
  diff -= update;
}
```
Not sure if Eigen provides a better interface that would also allow to avoid copy operations, and how much of a performance difference this actually makes.

Reply:
The *= operator does not exist for the sparse matrix type, otherwise this is implemented.